src/Plugin/formatter/Formatter.php

```diff
@@ -81,4 +84,104 @@ protected static function isIterable($input) {
     return is_array($input) || $input instanceof \Traversable || $input instanceof \stdClass;
   }
 
+  /**
+   * Checks if the passed in data to be rendered can be cached.
+   *
+   * @param mixed $data
+   *   The data to be prepared and rendered.
+   *
+   * @return bool
+   *   TRUE if the data can be cached.
+   */
+  protected function isCacheEnabled($data) {
+    // We are only caching field collections, but you could cache at different
+    // layers too.
+    if (!$data instanceof ResourceFieldCollectionInterface) {
+      return FALSE;
+    }
+    if (!$context = $data->getContext()) {
+      return FALSE;
+    }
+    return !empty($context['cache_fragments']);
+  }
+
+  /**
+   * Gets the cached computed value for the fields to be rendered.
+   *
+   * @param mixed $data
+   *   The data to be rendered.
+   *
+   * @return mixed
+   *   The cached data.
+   */
+  protected function getCachedData($data) {
+    if (!$render_cache = $this->createCacheController($data)) {
+      return NULL;
+    }
+    return $render_cache->get();
+  }
+
+  /**
+   * Gets the cached computed value for the fields to be rendered.
+   *
+   * @param mixed $data
+   *   The data to be rendered.
+   * @param mixed $output
+   *   The rendered data to output.
+   */
+  protected function setCachedData($data, $output) {
+    if (!$render_cache = $this->createCacheController($data)) {
+      return;
+    }
+    $render_cache->set($output);
+  }
+
+  /**
+   * Gets a cache controller based on the data to be rendered.
+   *
+   * @param mixed $data
+   *   The data to be rendered.
+   *
+   * @return RenderCacheInterface
+   *   The cache controller.
+   */
+  protected function createCacheController($data) {
+    $context = $data->getContext();
```

Comment:
> use isCacheEnabled?

We can't use it since `isCacheEnabled` only applies to `ResourceFieldCollections`.

> Also, do we really need to do this extra check - wasn't it in fact already tested in isCacheEnabled?

Just making sure, in case we want to use this helper method in other contexts.